docs/design/copy-info.md

```diff
@@ -0,0 +1,157 @@
+ignore copy info entirely until they are ready to implement it.
+
+### Read API
+
+Copy information will be served by a single new Backend trait method:
+
+```rust
+/// An individual copy source.
+pub struct CopySource {
+    /// The source path a target was copied from.
+    /// It is not requires that the source path is different than the target
+    /// path. A custom backend may choose to represent 'rollbacks' as copies
+    /// from a file unto itself, from a specific prior commit.
+    path: RepoPathBuf,
+    /// The source commit the target was copied from. If not specified, then the
+    /// parent of the target commit is the source commit.
+    id: Option<CommitId>,
+    // TODO: Record the FileId here?
+}
+
+/// An individual copy event, from file A -> B.
+pub struct CopyRecord {
+    /// The destination of the copy, B.
+    target: RepoPathBuf,
+    /// The CommitId where the copy took place.
+    id: CommitId,
+    /// The source of the copy, A.
+    source: CopySource,
+}
+
+/// Backend options for fetching copy records.
+pub struct CopyRecordOpts {
+    /// If true, follow transitive copy records. That is, if file A is copied to
+    /// B and then to C, a request for copy records of file C should also return
+    /// the copy from A to B.
+    transitive: bool
+    // TODO: Probably something for git similarity detection
+}
+
+pub trait Backend {
+    /// Get all copy records for `paths` in the dag range `roots..heads`.
+    fn get_copy_records(&self, paths: &[RepoPathBuf], heads: &[CommitId], roots: &[CommitId]) -> BackendResult<Box<dyn Iterator<Item = BackendResult<CopyRecord>> + '_>>;
+}
+```
+
+Obtaining copy records for a single commit requires first computing the files
+list for that commit, then calling get_copy_records with `heads = [id]` and
+`roots = parents()`. This enables commands like `jj diff` to produce better
+diffs that take copy sources into account.
+
+### Write API
+
+Backends that support explicit copy records at the commit level will do so
+through a new field on `backend::Commit` objects:
+
+```rust
+pub struct Commit {
+    ...
+    copies: Option<HashMap<RepoPathBuf, CopySource>>,
+}
+
+pub trait Backend {
+    /// Whether this backend supports explicit copy records on write.
+    fn supports_explicit_copies(&self) -> bool;
+}
+```
+
+This field will be write-only. Backends may choose to fill it on read, but have
+no obligation to do so, as clients are expected to use the `get_copy_records()`
+API to get copy info instead. A backend which does not support explicit copy
+info may choose to silently ignore it or return an error.
+
+This API will enable the creation of new `jj` commands for recording copies:
+
+```shell
+jj cp $SRC $DEST [OPTIONS]
+jj mv $SRC $DEST [OPTIONS]
+```
+
+These commands will perform the filesystem operations associated with
+copies/moves, as well as editing the working copy commit to include records of
+these copies.
```

Comment:
We generally perform operations on commits instead and then reflect those changes in the file system. I think that would work here too, right?

I think the commands should work on any commit (i.e. take `-r` arguments). We'll probably also want `--after` flags like `hg cp/mv` has for indicating that an existing file should be considered a copy/rename (without actually changing its contents).